README.md
# THIS PROJECT IS FORK FROM [nordnet/cordova-hot-code-push](https://github.com/nordnet/cordova-hot-code-push)

Because the origin project is stopped maintenance. We forked and do bug fix on this. PRs and ISSUEs welcome.

# Cordova Hot Code Push Plugin

This plugin provides functionality to perform automatic updates of the web based content in your application. Basically, everything that is stored in `www` folder of your Cordova project can be updated using this plugin.

When you publish your application on the store - you pack in it all your web content: html files, JavaScript code, images and so on. There are two ways how you can update it:

1. Publish new version of the app on the store. But it takes time, especially with the App Store.
2. Sacrifice the offline feature and load all the pages online. But as soon as Internet connection goes down - application won't work.

This plugin is intended to fix all that. When user starts the app for the first time - it copies all the web files onto the external storage. From this moment all pages are loaded from the external folder and not from the packed bundle. On every launch plugin connects to your server (with optional authentication, see fetchUpdate() below) and checks if the new version of web project is available for download. If so - it loads it on the device and installs on the next launch.

As a result, your application receives updates of the web content as soon as possible, and still can work in offline mode. Also, plugin allows you to specify dependency between the web release and the native version to make sure, that new release will work on the older versions of the application.

**Is it fine with App Store?** Yes, it is... as long as your content corresponds to what application is intended for and you don't ask user to click some button to update the web content. For more details please refer to [this wiki page](https://github.com/snipking/cordova-hot-code-push/wiki/App-Store-FAQ).

## Supported platforms

- Android 4.0.0 or above.
- iOS 7.0 or above. Xcode 7 is required.

### Installation

This requires cordova 5.0+ (current stable 1.5.8)

```sh
cordova plugin add cordova-hot-code-push-plugin-renew
```

It is also possible to install via repo url directly (__unstable__)
```sh
cordova plugin add https://github.com/snipking/cordova-hot-code-push.git
```

At the end of the installation plugin will recommend you to install [Cordova Hot Code Push CLI client](https://github.com/snipking/cordova-hot-code-push-cli). This client will help you to:
- easily generate necessary configuration files;
- launch local server to listen for any changes in the web project and deploy new version immediately on the app.

Of course, you can use this plugin without the CLI client, but it will make your life easier.

### Quick start guide

In this guide we will show how quickly you can test this plugin and start using it for development. For that we will install [development add-on](https://github.com/snipking/cordova-hot-code-push/wiki/Local-Development-Plugin).

1. Create new Cordova project using command line interface and add iOS/Android platforms:

  ```sh
  cordova create TestProject com.example.testproject TestProject
  cd ./TestProject
  cordova platform add android
  cordova platform add ios
  ```
  Or use the existing one.

2. Add plugin:

  ```sh
  cordova plugin add cordova-hot-code-push-plugin
  ```

3. Add plugin for local development:

  ```sh
  cordova plugin add cordova-hot-code-push-local-dev-addon
  ```

4. Install Cordova Hot Code Push CLI client:

  ```sh
  npm install -g cordova-hot-code-push-cli
  ```

5. Start local server by executing:

  ```sh
  cordova-hcp server
  ```

  As a result you will see something like this:
  ```
  Running server
  Checking:  /Cordova/TestProject/www
  local_url http://localhost:31284
  Warning: .chcpignore does not exist.
  Build 2015.09.02-10.17.48 created in /Cordova/TestProject/www
  cordova-hcp local server available at: http://localhost:31284
  cordova-hcp public server available at: https://5027caf9.ngrok.com
  ```

6. Open new console window, go to the project root and launch the app:

  ```sh
  cordova run
  ```

  Wait until application is launched for both platforms.

7. Now open your `index.html` page in `www` folder of the `TestProject`, change something in it and save. In a few seconds you will see updated page on the launched devices (emulators).

From this point you can do local development, where all the changes are uploaded on the devices without the need to restart applications on every change you made.

For production build do not forget to add the following to your `config.xml` file as it is a required property. Checkout the [wiki](https://github.com/snipking/cordova-hot-code-push/wiki/Cordova-config-preferences) for more information:

```xml
<chcp>
    <config-file url="https://5027caf9.ngrok.com/chcp.json"/>
</chcp>
```

### Documentation

All documentation can be found in details in our [Wiki on GitHub](https://github.com/snipking/cordova-hot-code-push/wiki).

If you have some questions/problems/suggestions - don't hesitate to post a [thread](https://github.com/snipking/cordova-hot-code-push/issues). If it's an actual issue - please, follow [this guide](https://github.com/snipking/cordova-hot-code-push/wiki/Issue-creation-guide) on how to do that properly.


CHANGELOG.md
# Change Log

## 1.5.9 (2020-07-20)

- Change `xcode` module require to support cordova 9

## 1.5.8 (2019-03-29)

- [Issue #1](https://github.com/snipking/cordova-hot-code-push/issues/1). Fixed by change plugin id from `cordova-hot-code-push-plugin` to `cordova-hot-code-push-plugin-renew`.

## 1.5.7 (2019-01-09)

- Support version update with iOS install. When old APK install with `1.0.0` and opened. Install new APK contains `1.0.1` and open will trigger version check. If new APK contains asset version number bigger than the old one. The new asset will install to local dir. Currently supported version number like `2018.12.19-15.26.06` and `1.0.0`.

## 1.5.6 (2018-12-19)

- Support version update with android APK install. When old APK install with `1.0.0` and opened. Install new APK contains `1.0.1` and open will trigger version check. If new APK contains asset version number bigger than the old one. The new asset will install to local dir. Currently supported version number like `2018.12.19-15.26.06` and `1.0.0`.

## 1.5.5 (2018-12-16)

- Version Number Up

## 1.5.4 (2018-12-14)

- Fix `chcp_updateLoadFailed` not fire on iOS device when no connection avaliable.
- Add `chcp_manifestDiffComplete` and `chcp_updateDownloadProgressEvent` event to indicate update download progress. [More info](https://github.com/snipking/cordova-hot-code-push/wiki/Listen-for-update-events)

## 1.5.3 (2016-05-27)

**Bug fixes:**

- [Issue #295](https://github.com/nordnet/cordova-hot-code-push/issues/295). Fixed build issues with Cordova 7 on iOS. Merged [pull request #296](https://github.com/nordnet/cordova-hot-code-push/pull/296), thanks to [@chauthai](https://github.com/chauthai).

**Documentation:**

- Merged [pull request #281](https://github.com/nordnet/cordova-hot-code-push/pull/281). Thanks to [@cadesalaberry](https://github.com/cadesalaberry).

## 1.5.2 (2016-11-06)

**Bug fixes:**

- [Issue #213](https://github.com/nordnet/cordova-hot-code-push/issues/213). Large sync lead to timeout on iOS. Merged [pull request #217](https://github.com/nordnet/cordova-hot-code-push/pull/217), thanks to [@bbreijer](https://github.com/bbreijer). Now iOS version should be more memory efficient when loading large set of files. It also can now download large files.
- Merged [pull request #218](https://github.com/nordnet/cordova-hot-code-push/pull/218). Fixes build failure on the after_prepare hook. Thanks to [@warent](https://github.com/warent).
- [Issue #228](https://github.com/nordnet/cordova-hot-code-push/issues/228). Added missing error code `NEW_APPLICATION_CONFIG_IS_INVALID` to JS API.
- [Issue #229](https://github.com/nordnet/cordova-hot-code-push/issues/229). Thanks to [@zwwhappy](https://github.com/zwwhappy) for [pull request #225](https://github.com/nordnet/cordova-hot-code-push/pull/225), and [@sfnt](https://github.com/sfnt) for [pull request #231](https://github.com/nordnet/cordova-hot-code-push/pull/231).

## 1.5.1 (2016-09-06)

**Bug fixes:**

- Merged [pull request #205](https://github.com/nordnet/cordova-hot-code-push/pull/205). Fixes Android app crashes, when `content_url` is missing in server's `chcp.json` config. Thanks to [@Mikey1982](https://github.com/Mikey1982).
- [Issue #208](https://github.com/nordnet/cordova-hot-code-push/issues/208). In some cases `plugins/ios.json` file couldn't be found by the hook, that checks, if `cordova-plugin-wkwebview-engine` plugin is installed in the project. Now hook will check for the plugin's folder directly.

## 1.5.0 (2016-08-30)

**Bug fixes:**

- [Issue #197](https://github.com/nordnet/cordova-hot-code-push/issues/197). Android can now update files with spaces in their names.

**Improvements:**

- Merged [pull request #204](https://github.com/nordnet/cordova-hot-code-push/pull/204). Initial installation of the assets on the external storage on Android is now much faster. Thanks to [@Mikey1982](https://github.com/Mikey1982).
- Added new JS method: `getVersionInfo`. Based on [pull request #170](https://github.com/nordnet/cordova-hot-code-push/pull/170) from [@Manduro](https://github.com/Manduro) - thank you. For documentation, please, refer to [wikki page](https://github.com/nordnet/cordova-hot-code-push/wiki/Get-version-information).

## 1.4.0 (2016-06-21)

**Bug fixes:**

- [Issue #155](https://github.com/nordnet/cordova-hot-code-push/issues/155). Android app should not crash if server has a bad `chcp.json` file.

**Improvements:**

- [Issue #153](https://github.com/nordnet/cordova-hot-code-push/issues/153). You can now pass `chcp.json` file url into `fetchUpdate` method on JS side. Also, you can provide additional HTTP headers to the request. For example, authorization info. These headers will be used for loading configuration files and updated/changed files from your server. Thanks to [@davidovich](https://github.com/davidovich) for [pull request #150](https://github.com/nordnet/cordova-hot-code-push/pull/150).
- [Issue #99](https://github.com/nordnet/cordova-hot-code-push/issues/99). iOS build version can now be a string. So, changing it from 1.0.0 to 1.0.1 will trigger reinstallation of `www` folder. Before that it had to be an integer.
- [Issue #160](https://github.com/nordnet/cordova-hot-code-push/issues/160). Old releases are now gets removed after update installation. Previously cleanup was performed only on application restart.

## 1.3.2 (2016-04-27)

**Bug fixes:**

- [Issue #137](https://github.com/nordnet/cordova-hot-code-push/issues/137). Fixed installation failure on Windows. Thanks to [@oxiao](https://github.com/oxiao) for providing the fix.

## 1.3.1 (2016-04-25)

**Bug fixes:**

- [Issue #139](https://github.com/nordnet/cordova-hot-code-push/issues/139). If you don't have `cordova-hcp` client - installation will not fail.
- Fixed typo in `before_plugin_install` hook. If installation failed - it will now throw a correct error.

**Enhancements:**


www/chcp.js


var exec = require('cordova/exec'),
  channel = require('cordova/channel'),

  // Reference name for the plugin
  PLUGIN_NAME = 'HotCodePush',

  // Plugin methods on the native side that can be called from JavaScript
  pluginNativeMethod = {
    INITIALIZE: 'jsInitPlugin',
    FETCH_UPDATE: 'jsFetchUpdate',
    INSTALL_UPDATE: 'jsInstallUpdate',
    CONFIGURE: 'jsConfigure',
    REQUEST_APP_UPDATE: 'jsRequestAppUpdate',
    IS_UPDATE_AVAILABLE_FOR_INSTALLATION: 'jsIsUpdateAvailableForInstallation',
    GET_INFO: 'jsGetVersionInfo'
  };

// Called when Cordova is ready for work.
// Here we will send default callback to the native side through which it will send to us different events.
channel.onCordovaReady.subscribe(function() {
  ensureCustomEventExists();
  exec(nativeCallback, null, PLUGIN_NAME, pluginNativeMethod.INITIALIZE, []);
});

/**
 * Method is called when native side sends us different events.
 * Those events can be about update download/installation process.
 *
 * @param {String} msg - JSON formatted string with call arguments
 */
function nativeCallback(msg) {
  // parse call arguments
  var resultObj = processMessageFromNative(msg);
  if (resultObj.action == null) {
    console.log('Action is not provided, skipping');
    return;
  }

  broadcastEventFromNative(resultObj);
}

/**
 * Parse arguments that were sent from the native side.
 * Arguments are a JSON string of type:
 * { action: "action identifier", error: {...error data...}, data: {...some additional data..} }
 * Some parameters may not exist, but the resulting object will have them all.
 *
 * @param {String} msg - arguments as JSON string
 * @return {Object} parsed string
 */
function processMessageFromNative(msg) {
  var errorContent = null,
    dataContent = null,
    actionId = null;

  try {
    var resultObj = JSON.parse(msg);
    if (resultObj.hasOwnProperty('error')) {
      errorContent = resultObj.error;
    }
    if (resultObj.hasOwnProperty('data')) {
      dataContent = resultObj.data;
    }
    if (resultObj.hasOwnProperty('action')) {
      actionId = resultObj.action;
    }
  } catch (err) {}

  return {
    action: actionId,
    error: errorContent,
    data: dataContent
  };
}

function callNativeMethod(methodName, options, callback) {
  var innerCallback = function(msg) {
    var resultObj = processMessageFromNative(msg);
    if (callback !== undefined && callback != null) {
      callback(resultObj.error, resultObj.data);
    }
  };

  var sendArgs = [];
  if (options !== null && options !== undefined) {
    sendArgs.push(options);
  }

  exec(innerCallback, null, PLUGIN_NAME, methodName, sendArgs);
}

// region Update/Install events

/*
 * Polyfill for adding CustomEvent which may not exist on older versions of Android.
 * See https://developer.mozilla.org/fr/docs/Web/API/CustomEvent for more details.
 */
function ensureCustomEventExists() {
  // Create only if it doesn't exist
  if (window.CustomEvent) {
    return;
  }

  var CustomEvent = function(event, params) {
    params = params || {
      bubbles: false,
      cancelable: false,
      detail: undefined
    };
    var evt = document.createEvent('CustomEvent');
    evt.initCustomEvent(event, params.bubbles, params.cancelable, params.detail);
    return evt;
  };

  CustomEvent.prototype = window.Event.prototype;
  window.CustomEvent = CustomEvent;
}

/**
 * Broadcast event that was received from the native side.
 *
 * @param {Object} arguments, received from the native side
 */
function broadcastEventFromNative(nativeMessage) {
  var params = {};
  if (nativeMessage.error != null) {
    params.error = nativeMessage.error;
  }

  if (nativeMessage.data != null) {
    params.data = nativeMessage.data;
  }

  var chcpEvent = new CustomEvent(nativeMessage.action, {
    'detail': params
  });
  document.dispatchEvent(chcpEvent);
}

// endregion

/**
 * Public module of the plugin.
 * May be used by developers to send commands to the plugin.
 */
var chcp = {

  // Plugin error codes
  error: {
    NOTHING_TO_INSTALL: 1,
    NOTHING_TO_UPDATE: 2,

    FAILED_TO_DOWNLOAD_APPLICATION_CONFIG: -1,
    APPLICATION_BUILD_VERSION_TOO_LOW: -2,
    FAILED_TO_DOWNLOAD_CONTENT_MANIFEST: -3,
    FAILED_TO_DOWNLOAD_UPDATE_FILES: -4,
    FAILED_TO_MOVE_LOADED_FILES_TO_INSTALLATION_FOLDER: -5,
    UPDATE_IS_INVALID: -6,
    FAILED_TO_COPY_FILES_FROM_PREVIOUS_RELEASE: -7,
    FAILED_TO_COPY_NEW_CONTENT_FILES: -8,
    LOCAL_VERSION_OF_APPLICATION_CONFIG_NOT_FOUND: -9,
    LOCAL_VERSION_OF_MANIFEST_NOT_FOUND: -10,
    LOADED_VERSION_OF_APPLICATION_CONFIG_NOT_FOUND: -11,
    LOADED_VERSION_OF_MANIFEST_NOT_FOUND: -12,
    FAILED_TO_INSTALL_ASSETS_ON_EXTERNAL_STORAGE: -13,
    CANT_INSTALL_WHILE_DOWNLOAD_IN_PROGRESS: -14,
    CANT_DOWNLOAD_UPDATE_WHILE_INSTALLATION_IN_PROGRESS: -15,
    INSTALLATION_ALREADY_IN_PROGRESS: -16,
    DOWNLOAD_ALREADY_IN_PROGRESS: -17,
    ASSETS_FOLDER_IN_NOT_YET_INSTALLED: -18,
    NEW_APPLICATION_CONFIG_IS_INVALID: -19
  },

  // Plugin events
  event: {
    BEFORE_ASSETS_INSTALLATION: 'chcp_beforeAssetsInstalledOnExternalStorage',
    ASSETS_INSTALLATION_FAILED: 'chcp_assetsInstallationError',
    ASSETS_INSTALLED: 'chcp_assetsInstalledOnExternalStorage',

    NOTHING_TO_UPDATE: 'chcp_nothingToUpdate',
    UPDATE_LOAD_FAILED: 'chcp_updateLoadFailed',
    UPDATE_IS_READY_TO_INSTALL: 'chcp_updateIsReadyToInstall',

    BEFORE_UPDATE_INSTALLATION: 'chcp_beforeInstall',
    UPDATE_INSTALLATION_FAILED: 'chcp_updateInstallFailed',
    UPDATE_INSTALLED: 'chcp_updateInstalled',
    NOTHING_TO_INSTALL: 'chcp_nothingToInstall',
    MANIFEST_DIFF_COMPLETE: 'chcp_manifestDiffComplete',
    UPDATE_DOWNLOAD_PROGRESS: 'chcp_updateDownloadProgressEvent',
    UPDATE_INSTALL_PROGRESS: 'chcp_updateInstallProgressEvent'
  },

  /**
   * DEPRECATED! WILL BE REMOVED EVENTUALLY!
   *
   * If you want to set config-url - use chcp.fetchUpdate(callback, options).
   * If you want to set auto-download/auto-install preference - do it in config.xml instead of this method.
   *
   * Set plugin options.
   * Options are send to the native side.
   * As soon as they are processed - callback is called.
   *
   * @param {Object} options - options to set
   * @param {Callback(error)} callback - callback to call when options are set
   */
  configure: function(options, callback) {
    if (options === undefined || options == null) {
      return;
    }

    callNativeMethod(pluginNativeMethod.CONFIGURE, options, callback);
  },

  /**
   * Show dialog with the request to update application through the Store (App Store or Google Play).
   * For that purpose you can use any other cordova library, this is just a small helper method.
   *
   * @param {String} message - message to show in the dialog
   * @param {Callback()} onStoreOpenCallback - called when user redirects to the Store
   * @param {Callback()} onUserDismissedDialogCallback - called when user declines to go to the Store
   */
  requestApplicationUpdate: function(message, onStoreOpenCallback, onUserDismissedDialogCallback) {
    if (message == undefined || message.length == 0) {
      return;
    }

    var onSuccessInnerCallback = function(msg) {
      if (onStoreOpenCallback) {
        onStoreOpenCallback();
      }
    };

    var onFailureInnerCallback = function(msg) {
      if (onUserDismissedDialogCallback) {
        onUserDismissedDialogCallback();
      }
    };

    exec(onSuccessInnerCallback, onFailureInnerCallback, PLUGIN_NAME, pluginNativeMethod.REQUEST_APP_UPDATE, [message]);
  },

  /**
   * Check if any new content is available on the server and download it.
   * Usually this is done automatically by the plugin, but can be triggered at any time from the web page.
   *
   * @param {Callback(error, data)} callback - called when native side finished update process
   * @param {Object} options - additional options, such as "config-url" and additional http headers.
   */
  fetchUpdate: function(callback, options) {
    callNativeMethod(pluginNativeMethod.FETCH_UPDATE, options, callback);
  },

  /**
   * Install update if there is anything to install.
   *
   * @param {Callback(error)} callback - called when native side finishes installation process
   */
  installUpdate: function(callback) {
    callNativeMethod(pluginNativeMethod.INSTALL_UPDATE, null, callback);
  },

  /**
   * Check if update was loaded and ready to be installed.
   * If update was loaded and can be installed - "data" property of the callback will contain the name of the current version and the name of the new
   * version.
   * If not - "error" will contain code chcp.error.NOTHING_TO_INSTALL.
   *
   * @param {Callback(error, data)} callback - called, when information is retrieved from the native side.
   */
  isUpdateAvailableForInstallation: function(callback) {
    callNativeMethod(pluginNativeMethod.IS_UPDATE_AVAILABLE_FOR_INSTALLATION, null, callback);
  },

  /**
   * Get information about the current version like current release version, app build version and so on.
   * The "data" property of the callback will contain all the information.
   *
   * @param {Callback(error, data)} callback - called, when information is retrieved from the native side.
   */
  getVersionInfo: function(callback) {
    callNativeMethod(pluginNativeMethod.GET_INFO, null, callback);
  }
};

module.exports = chcp;
